teps/0044-decouple-task-composition-from-scheduling.md

```diff
@@ -0,0 +1,954 @@
+  [local execution](https://github.com/tektoncd/pipeline/issues/235)
+* Users don't have to think about how the Pipeline will eb executed
+* Pipelines can be used without updates
+
+Cons:
+* Only cluster administrators will be able to control this scheduling, there will be no runtime or authoring time
+  flexibility
+* Executing a pipeline in a pod will require significantly re-architecting our graph logic so it can execute outside
+  the controller and has a lot of gotchas we'll need to iron out (see
+  [https://hackmd.io/@vdemeester/SkPFtAQXd](https://hackmd.io/@vdemeester/SkPFtAQXd) for some more brainstorming)
+
+### Within the Task
+
+In this option we ignore the "not at Task authoring time" requirement and we allow for Tasks to contain other Tasks.
+
+This is similar to [TEP-0054](https://github.com/tektoncd/community/pull/369) which proposes this via the Task spec
+in a Pipeline, but does not (yet) propose it for Tasks outside of Pipelines.
+
+For example:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Task
+metadata:
+  name: build-test-upload
+spec:
+  workspaces:
+  - name: source
+    mountPath: /workspace/source/go/src/github.com/GoogleContainerTools/skaffold
+  steps:
+  - name: get-source
+    uses: git-clone
+    params:
+      url: $(params.url)
+    workspaces:
+    - name: source
+      workspace: source
+  - name: run-tests
+    image: golang
+    workingDir: $(workspaces.source.path)
+    script: |
+      go test <stuff here>
+  - name: upload-results
+    uses: gcs-upload
+```
+
+Pros:
+* Doesn't require many new concepts
+
+Cons:
+* Can create confusing chains of nested Tasks (Task A can contain Task B which can Contain Task C...)
+* Requires creating new Tasks to leverage the reuse (maybe embedded specs negate this?)
+* Doesn't help with parallel use cases
+
+### Remove distinction between Tasks and Pipelines
+
+In this version, we try to combine Tasks and Pipelines into one thing; e.g. by getting rid of Pipelines and adding all
+the features they have to Tasks, and by giving Tasks the features that Pipelines have which they do not have.
+
+Things Tasks can do that Pipelines can't:
+* Sidecars
+* Refer to images (including args to images like script, command, args, env....)
+
+Things Pipelines can do that Tasks can't:
+* Create DAGs, including running in parallel
+* Finally
+* When expressions
+
+For example, say our new thing is called a Foobar:
+
+```yaml
+kind: Foobar
+metadata:
+  name: git-clone
+spec:
+ workspaces:
+  - name: source-code
+ foobars:
+ - name: get-source
+   steps: # or maybe each FooBar can only have 1 step and we need to use runAfter / dependencies to indicate ordering?
+    - name: clone
+      image: gcr.io/tekton-releases/github.com/tektoncd/pipeline/cmd/git-init:v0.21.0
+      script: <script here>
+   workspaces:
+   - name: source-code
+     workspace: source-code
+ finally:
+ # since we merged these concepts, any Foobar can have a finally
```

Comment:
at which level(s) would whenexpressions fit in this example? would it guard `Foobar`, each of the `foobars` or `steps`? (recently saw a user asking to guard steps with whenexpressions)